base/compiler/typeinfer.jl

```diff
@@ -849,16 +856,18 @@ function typeinf_edge(interp::AbstractInterpreter, method::Method, @nospecialize
         end
         typeinf(interp, frame)
         update_valid_age!(caller, frame.valid_worlds)
-        edge = is_inferred(frame) ? mi : nothing
-        return EdgeCallResult(frame.bestguess, edge, frame.ipo_effects) # effects are adjusted already within `finish`
+        isinferred = is_inferred(frame)
+        edge = isinferred ? mi : nothing
+        effects = isinferred ? frame.ipo_effects : adjust_effects(Effects(), method) # effects are adjusted already within `finish` for ipo_effects
```

Comment:
`Effects()` might be too wide and it causes external `AbsInt` with overlay method table to see inference regression for `typejoin(a, b)` again and leads to the test failure in test/compiler/AbstractInterpreter.jl (xref: #48097).

Maybe we should employ a similar technology as what we do for type convergence?

Reply:
Except that these are often not converged for types either, since know that is expensive and usually not beneficial, and which is how I encountered this. I think you might need to go back and a special case for it that slows down inference when it is using overlay tables.

It seems related to this regression too, since it is now trying abstract inference on typejoin, even though it should not:
```
julia> length(which(typejoin,(Any,Any)).specializations)
1 # v1.6
90 # v1.9
163 # v1.10
```
I added a commit to fix that, using your recent work on `@nospecializeinfer`